### ISSUE TITLE
Spec: cast values for be_nil and be_a expectations

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#8369]
A way to solve [this forum request](https://forum.crystal-lang.org/t/var-should-not-be-nil-does-not-eliminate-var-of-nil-from-subsequent-type-inference/1165)

Currently, when validating the type of a value in spec, we need to restrict the type afterwards in order to call restricted methods on the sample:

```crystal
index = "foo".index('o')
index.should_not be_nil
foo[index.not_nil!].should eq('o')
```

The `not_nil!` is needed because the compiler has no way to know that the `should_not be_nil` means that `index` will not be `nil`.

The idea is to let `be_nil` and `be_a` expectations return the value with appropriate type restriction:
```crystal
index = "foo".index('o')
index = index.should_not be_nil
foo[index].should eq('o')
```

Basically, the result of `obj.should_not be_nil` will return `obj` as a non-nil object.

The same applies for `obj.should be_a(T)` and `obj.should_not be_a(T)`.

An implementation was submitted in #8240, but it should be overhauled.
